test/serialization/serialization-test.cpp

```diff
@@ -0,0 +1,165 @@
+    std::istringstream in(buffer);
+    found::DataFile parsed = deserialize(in);
+
+    // Validate the parsed header fields
+    ASSERT_EQ(parsed.header.version, 1U);
+    ASSERT_EQ(parsed.header.num_positions, 2U);
+
+    // Validate the relative attitude fields
+    ASSERT_EQ(parsed.relative_attitude.roll, data.relative_attitude.roll);
+    ASSERT_EQ(parsed.relative_attitude.ra, data.relative_attitude.ra);
+    ASSERT_EQ(parsed.relative_attitude.de, data.relative_attitude.de);
+
+    // Validate the positions
+    ASSERT_EQ(parsed.positions.size(), 2U);
+    EXPECT_EQ(parsed.positions[0].timestamp, loc1.timestamp);
+    EXPECT_EQ(parsed.positions[0].position.x, loc1.position.x);
+    EXPECT_EQ(parsed.positions[1].timestamp, loc2.timestamp);
+    EXPECT_EQ(parsed.positions[1].position.y, loc2.position.y);
+}
+
+TEST_F(SerializationTest, CorruptedPositionDeserialization) {
+    // Create a header that claims to have one position
+    std::ostringstream out;
+    DataFileHeader header;
+    memcpy(header.magic, "FOUN", 4);
+    header.version = 1;
+    header.num_positions = 1;  // Claims one record
+
+    // Calculate and assign the CRC
+    header.crc = found::calculateCRC32(&header, sizeof(header) - sizeof(header.crc));
+
+    // Convert the header fields to network byte order (big endian)
+    header.version = htonl(header.version);
+    header.num_positions = htonl(header.num_positions);
+    header.crc = htonl(header.crc);  // Convert CRC as well
+
+    // Write the header to the output stream
+    out.write(reinterpret_cast<const char*>(&header), sizeof(header));
+
+    // Write only 3 bytes of position data (incomplete LocationRecord)
+    out.write("bad", 3);  // Simulate a corrupted or truncated file
+
+    std::string buffer = out.str();
+    std::istringstream in(buffer);
+
+    // Attempt to deserialize and expect a failure due to corrupted input
+    EXPECT_THROW({
+        found::deserialize(in);
+    }, std::ios_base::failure);
+}
+
+TEST_F(SerializationTest, MagicNumberMismatch) {
+    // Create a header with an incorrect magic number
+    std::ostringstream out;
+    DataFileHeader header;
+    memcpy(header.magic, "FAIL", 4);  // Wrong magic value
+    header.version = 1;
+    header.num_positions = 0;
+
+    // Compute CRC based on the incorrect magic
+    header.crc = found::calculateCRC32(&header, sizeof(header) - sizeof(header.crc));
+
+    // Convert to network byte order
+    header.version = htonl(header.version);
+    header.num_positions = htonl(header.num_positions);
+    header.crc = htonl(header.crc);
+
+    // Write the faulty header to stream
+    out.write(reinterpret_cast<const char*>(&header), sizeof(header));
+
+    std::string buffer = out.str();
+    std::istringstream in(buffer);
+
+    // The CRC might pass, but the magic number should trigger a validation failure
+    EXPECT_THROW({
+        readHeader(in);
+    }, std::runtime_error);
+}
```

Comment:
I think I'd like to see a bit more test coverage here, please add tests for the following cases:
1. No positions (this is actually common)
2. No relative orientation (Think about what your code should do in this case, should we tolerate such a case?)
3. Nothing else except a header (Again, do you want to tolerate this case?)